src/order-radix.c

```diff
@@ -237,15 +240,87 @@ SEXP vctrs_order_locs(SEXP x,
 
 static
 SEXP vec_order_locs(SEXP x, SEXP direction, SEXP na_value, bool nan_distinct, SEXP chr_transform) {
-  return vec_order_impl(x, direction, na_value, nan_distinct, chr_transform, true);
+  SEXP info = PROTECT(vec_order_info(x, direction, na_value, nan_distinct, chr_transform));
+
+  SEXP o = VECTOR_ELT(info, 0);
+  const int* p_o = INTEGER_RO(o);
+
+  SEXP sizes = VECTOR_ELT(info, 1);
+  const int* p_sizes = INTEGER_RO(sizes);
+
+  r_ssize n_groups = r_length(sizes);
+
+  SEXP loc = PROTECT(Rf_allocVector(VECSXP, n_groups));
+
+  SEXP key_loc = PROTECT(Rf_allocVector(INTSXP, n_groups));
+  int* p_key_loc = INTEGER(key_loc);
+
+  int start = 0;
+
+  for (r_ssize i = 0; i < n_groups; ++i) {
+    p_key_loc[i] = p_o[start];
+
+    const int size = p_sizes[i];
+
+    SEXP elt = Rf_allocVector(INTSXP, size);
+    SET_VECTOR_ELT(loc, i, elt);
+    int* p_elt = INTEGER(elt);
+
+    R_len_t k = 0;
+
+    for (int j = 0; j < size; ++j) {
+      p_elt[k] = p_o[start];
+      ++start;
+      ++k;
+    }
+  }
+
+  SEXP key = PROTECT(vec_slice(x, key_loc));
+
+  // Construct output data frame
```

Comment:
`r_alloc_data_frame()` might be useful https://github.com/r-lib/memtools/blob/75d41655/src/snapshot.c#L171-L175

Reply:
`key` has the generic type of `x`, so it isn't one of atomic `RTYPE_*` values, which might make usage of this difficult?